cdap-docs/admin-manual/source/operations/logging.rst

```diff
@@ -10,35 +10,106 @@ Logging and Monitoring
 
 Logs and Metrics
 ================
-CDAP collects logs and metrics for all of its internal services and and user applications.
+CDAP collects logs and metrics for all of its internal services and on user applications.
 Being able to view these details can be very helpful in debugging CDAP applications as
 well as analyzing their performance. CDAP gives access to its logs, metrics, and other
-monitoring information through RESTful APIs as well as a Java Client.
+monitoring information through RESTful APIs, the CDAP UI, as well as a Java Client.
 
-See the :ref:`Logging <http-restful-api-logging>`, :ref:`Metrics <http-restful-api-metrics>`, 
-and :ref:`Monitoring <http-restful-api-monitor>` HTTP RESTful APIs, the :ref:`Java Client
-<reference:java-client-api>`, and the :ref:`master-service-logging-configuration` for
-additional information.
+In Hadoop clusters, the programs running in containers generate individual log files as
+part of the container. As these files generally do not persist beyond the life of the
+container, they are volatile and not very useful for post-mortem diagnostics and
+troubleshooting, or for analyzing performance. As a single application can consist of
+multiple  programs distributed across the nodes of a cluster, the complete logs for the
+application may similarly be distributed.
 
+To address these issues, the CDAP log framework was designed:
 
-Log Location
-============
-The location of logs depends on the mode of CDAP and the Hadoop distribution.
+- to centralize the location of logs, so that the individual logs of an application's programs can be
+  merged into one;
 
-- With **Standalone CDAP** (the CDAP SDK), logs are located in ``<CDAP-SDK-HOME>/logs``.
+- to make logs persistent and available for later use and analysis; and
 
-- For **Distributed CDAP**, with the exception of Cloudera Manager, logs are located in
-  ``/var/log/cdap``. With Cloudera Manager installations, log files are located in
-  directories under ``/var/run/cloudera-scm-agent/process``.
+- to use a pluggable system that can be adapted for individual requirements at both the
+  application and cluster levels.
 
 
-Logging Messages
-================
+Logging Framework
+=================
+
+.. highlight:: console
+
+This diagram shows the steps CDAP follows with its logging framework:
+
+  1. Logs are collected from individual programs running in YARN containers.
+
+  #. YARN containers have their own local log files.
+
+  #. CDAP logs are published to Kafka topics.
+
+  #. The Kafka topics are consumed by the Log Saver Service. Each instance of the service
+     has one or more logging pipelines. Each pipeline buffers log messages in-memory and
+     sorts them based on their timestamp. Custom pipelines can be configured by adding an
+     XML file in a prescribed location.
```

Comment:
Log saver service is configured to read log messages for "logs.user-v2" kafka topic. 
Log saver instance can be scaled to process the kafka partitions in parallel, if needed. 
Log.saver reads the messages from kafka, buffers and sorts them in memory, before persisting to HDFS. These messages are written to files corresponding to the program the log messages belong to. 

Note: these files are configured to rotate based on time and size, they can be changed using these properties - log.pipeline.cdap.file.max.size.bytes and log.pipeline.cdap.file.max.lifetime.ms

The files are written with permission such that they are accessible only by the "cdap" user for security reasons.

Log.saver also emits metrics for each {info,debug,error} messages received. this helps users to monitor the amount of logs emitted by application,  application's programs, system services, etc by querying CDAP metrics system.